KAFKA-13456: Tighten KRaft config checks/constraints

rondagostino opened the pull request:
This patch tightens the configuration checks related to KRaft configs by adding the following constraints:

* `control.plane.listener.name` is confirmed to be empty in KRaft mode whenever a config object is created as opposed to later when the broker is given the config and tries to start.
* `controller.listener.names` is required to be empty for the non-KRaft (i.e. ZooKeeper) case.  A ZooKeeper-based cluster that sets this config will fail to restart until this config is removed.
* There must be no advertised listeners when running just a KRaft controller (i.e. when `process.roles=controller`).  This means neither `listeners` nor `advertised.listeners` (if the latter is explicitly defined) can contain a listener that does not also appear in `controller.listener.names`.
* When running a KRaft broker (i.e. when `process.roles=broker` or `process.roles=broker,controller`), advertised listeners (which was already checked to be non-empty via the check that the inter-broker listener appear there) must not include any listeners appearing in `controller.listener.names`.
* When running a KRaft controller (i.e. when `process.roles=controller` or `process.roles=broker,controller`) `controller.listener.names` must be non-empty and every one must appear in `listeners`
* When running just a KRaft broker (i.e. when `process.roles=broker`) `controller.listener.names` must be non-empty and none of them can appear in `listeners`.  This was indirectly checked previously, but the indirect checks did not catch all cases.
* When running just a KRaft broker we log a warning if more than one entry appears in `controller.listener.names` because only the first entry is used.

We also map the `CONTROLLER` listener name to the `PLAINTEXT` security protocol by default when using KRaft as a convenience.

This patch adds appropriate unit tests.  I have manually confirmed that our KRaft Quickstart config files in `config/kraft` remain valid (both for the combined and separate cases).  I also confirmed that KRaft system test configs remain valid (see sanity check test output below).

```
================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.8.1
session_id:       2021-11-23--001
run time:         6 minutes 52.763 seconds
tests run:        5
passed:           5
failed:           0
ignored:          0
================================================================================
test_id:    kafkatest.sanity_checks.test_bounce.TestBounce.test_simple_run.metadata_quorum=COLOCATED_KRAFT.quorum_size=3
status:     PASS
run time:   1 minute 56.752 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.sanity_checks.test_bounce.TestBounce.test_simple_run.metadata_quorum=REMOTE_KRAFT.quorum_size=3
status:     PASS
run time:   1 minute 49.306 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.sanity_checks.test_bounce.TestBounce.test_simple_run.metadata_quorum=COLOCATED_KRAFT.quorum_size=1
status:     PASS
run time:   59.592 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.sanity_checks.test_bounce.TestBounce.test_simple_run.metadata_quorum=REMOTE_KRAFT.quorum_size=1
status:     PASS
run time:   1 minute 9.941 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.sanity_checks.test_bounce.TestBounce.test_simple_run.metadata_quorum=ZK.quorum_size=1
status:     PASS
run time:   56.642 seconds
--------------------------------------------------------------------------------
``` 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rondagostino committed:
KAFKA-13456: controller.listener.names required in KRaft

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2007,8 +2007,13 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
       " Valid options are " + BrokerCompressionCodec.brokerCompressionOptions.mkString(","))
-    require(!processRoles.contains(ControllerRole) || controllerListeners.nonEmpty,
-      s"${KafkaConfig.ControllerListenerNamesProp} cannot be empty if the server has the controller role")
+    if (usesSelfManagedQuorum) {
```
A few additional validations we can do:

1. Ensure control plane listener is empty. We have this check in BrokerServer already, but we should check it here as well.
2. Verify port consistency between the controller listener and the quorum voter string. As you noted offline, the port gets duplicated, so we should be sure they match. 
3. We currently assume that the first listener in `controller.listener.names` is the one that matches the voter string. We should either require this to be the case, or change the logic in `RaftManager` to choose the right listener.

jsancio replied:
> 2. Verify port consistency between the controller listener and the quorum voter string. As you noted offline, the port gets duplicated, so we should be sure they match.

One complexity is when the controllers let the OS assign a port (`0`) to the listener. This feature is heavily used in the tests to allow multiple controllers to run on the same host.

hachikuji replied:
I think that's fine. When the test cases override `controller.quorum.voters`, they also override the controller listener. For example, see here: https://github.com/apache/kafka/blob/trunk/core/src/test/scala/integration/kafka/server/QuorumTestHarness.scala#L220. So it seems reasonable to validate port consistency in any case.

jsancio replied:
Okay. How about this algorithm/check:

If `process.roles=broker` then Kafka requires `controller.listener.names` is set to one name and doesn't allow the `listeners` to contain `controller.listener.names`. This is consistent since controller ports are only specified in `controller.quorum.voters`

When `process.roles=controller,*` then Kafka requires `controller.listener.names` is set and `listeners` contains `controller.listener.names`. Additionally check that the port in `controller.quorum.voters` for `node.id` matches the port in the `listeners` for one of `controller.listener.names`. Fix `RaftManager` to pick the right controller listener name based on the port of it's id in the voters.

hachikuji replied:
I'm a little unsure about requiring the broker have only one controller listener. Technically it's an incompatible change. Another way we could handle it is to log a warning that other specified ports will not be used. 

The second part sounds good to me.

rondagostino replied:
I'm in agreement with not allowing `listeners` to contain the `controller.listener.names` values(s).  I'll rework tghe PR to do that.

FYI, just as a sanity check, I confirmed that I was able to get a separate controller and broker working using our QuickStart files with the following changes:

**controller.properties**
```
listeners=SASL_PLAINTEXT://:9093
listener.name.sasl_plaintext.oauthbearer.sasl.jaas.config=org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule Required unsecuredLoginStringClaim_sub="broker";
sasl.enabled.mechanisms=OAUTHBEARER
sasl.mechanism.controller.protocol=OAUTHBEARER
controller.listener.names=SASL_PLAINTEXT
```

**broker.properties**
```
listener.security.protocol.map=CONTROLLER:SASL_PLAINTEXT,PLAINTEXT:PLAINTEXT,SSL:SSL,SASL_PLAINTEXT:SASL_PLAINTEXT,SASL_SSL:SASL_SSL
sasl.mechanism.controller.protocol=OAUTHBEARER
listener.name.controller.oauthbearer.sasl.jaas.config=org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule Required unsecuredLoginStringClaim_sub="broker";
```

I was able to do this both with requiring the additional listener on the broker (the initial version of this PR) and not (the way the code works without this PR).

So while the "listener" terminology is unfortunate in that the broker is not "listening" on it, I have confirmed that everything works as expected.

I'll work on this more tomorrow.

hachikuji replied:
@rondagostino How about the 3 additional validations I mentioned in the initial comment above?

rondagostino replied:
Pushed a commit that does the following:

- Ensure control plane listener is empty
- If process.roles=broker then don't allow the listeners to contain controller.listener.names
- When process.roles=controller,* then require controller.listener.names is set and listeners contains controller.listener.names

Still need to verify port consistency between the controller listener and the quorum voter string.  There is some subtlety regarding what the right thing to do is on a broker-only node vs one that has the controller node, which I will have to think about a bit.

rondagostino replied:
Ok, I pushed a commit that checks the consistency of the ports between the `controller.quorum.voters` and (`controller.listener.names`/`listeners`) configs.  This can only be done for the controller case since there is nothing in `listeners` for the broker-only case.

Currently both `RaftManager` and `RaftControllerNodeProvider` take `config.controllerListenerNames.head` to determine how to connect.

`RaftControllerNodeProvider` extends `ControllerNodeProvider` which exposes a single listener name/security protocol/SASL mechanism.  So we aren't going to make `RaftControllerNodeProvider` sophisticated about which security protocol/SASL mechanism to use based on voter ports -- it has to always use a single one.  This makes sense given that migrating communication to the ZK-based brokers (any one of which can become the controller) from one security protocol to another involves adding the second security protocol listener everywhere, migrating everybody over to using that second security protocol (e.g. via `inter.broker.security.protocol`), and then dropping the first listener -- so 3 rolls.  We would do something similar in KRaft: add the second controller listener on all controllers, migrate all brokers to use it (by making that listener the first one in the controller listener names list), then drop the original listener on the controllers.

It now becomes appropriate to discuss `RaftManager`.  `RaftManager` is used on all KRaft nodes, including any that are broker-only nodes, and in that broker-only case it is going to be unable to use ports in `controller.quorum.voters` to identify a security protocol because there is nothing in `listeners` to find based on a port number.  So it has to use `config.controllerListenerNames.head` in this case.

`RaftManager` could potentially be sophisticated about which security protocol/SASL mechanism to use based on voter ports in the case where we have the controller role -- there are listeners that it can identify based on port numbers.  But I don't think we would want to in this case, which generally corresponds to the above migration scenario.  I think the way that scenario would work in KRaft would be as follows:

1. Append a second entry to `controller.listener.names` to all nodes having the controller role; add a corresponding entry in `listeners`; and roll all controllers.  This has no effect other than adding the second active controller listener to each controller -- in particular, they won't hear anything yet.
2. Now that all controllers have both listeners, swap the order of `controller.listener.names` in all controller nodes; update `controller.quorum.voters` with the new port; and roll all controllers again.  Now the controllers are talking to each other on that new listener/security protocol, and any brokers running in combined mode are as well.
3. Set `controller.listener.names` to be the new listener name and update `controller.quorum.voters` on any broker-only nodes and roll them.  Now the broker-only nodes are also talking to the controllers over the new security protocol.
4. Eliminate the old entry from `controller.listener.names` on any controller node (it is now the second in the list at this point, and we are deleting it), and roll the controllers a third time.  We have completed the migration.

So it takes up to 4 rolls: 2 rolls of the controller nodes followed by one roll of any broker-only nodes (skip this if there are none) followed by a final roll of the controller nodes.

The point here is that we never have a situation where a running instance of a broker or a controller is talking to some controllers with one security protocol and others with another security protocol.  Every node always talks to all controllers over the same security protocol at any particular moment.

Assuming this all makes sense and we agree, I think this has the following implications:

* I added validation of port consistency for controllers by checking that the port existed in at least one controller listener.  I think this validation should be more strict: the port must match against the first controller listener listed.
* For broker-only nodes, `controller.listener.names` should only have a single entry.  Given the point above that this is technically an incompatible change, I think the suggestion to log a warning is appropriate.

Thoughts?

rondagostino replied:
I've defined the voter port validation to check that this controller's voter node port matches the first controller listener port.  We can't validate all controllers' voter ports because we run multiple controllers on the same host in tests, and in that case every controller must list on a different port despite using the same security protocol.

rondagostino replied:
I've removed voter port validation because an offline discussion pointed out that we can't validate the `controller.quorum.voters` ports (which are the ports that clients use and are semantically "advertised" ports even though the controller configuration doesn't list them in `advertised.listeners`) against the `listener` ports (which are semantically different then the ports that clients use).

rondagostino committed:
Initial pass at final requirements

rondagostino committed:
Validate consistency of voter and controller listener ports for controller case

jsancio commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2007,12 +2008,47 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+        val listenerNameValues = listeners.map(_.listenerName.value).toSet
+        require(controllerListenerNames.forall(cln => listenerNameValues.contains(cln)),
+          s"${KafkaConfig.ControllerListenerNamesProp} must only contain values appearing in the '${KafkaConfig.ListenersProp}' configuration when running the KRaft controller role")
+        RaftConfig.parseVoterConnections(quorumVoters).asScala.foreach { case (nodeId, addressSpec) =>
```
This checks the port of every voter. We should only check the port of this voter/controller. Instead of `foreach` maybe `get(nodeId)`? This must be `Some` because I think we have another validation check for the id of this controller is in the voter set.

rondagostino replied:
> We should only check the port of this voter/controller

I coded it that way first, but upon further thought I came to the conclusion that we should check every voter -- not just this one.  Based on my comment above, I believe this should actually require the port to match the first controller listener's port, and that further adds weight to checking every node.

rondagostino replied:
Hmm... what happens for the case of running multiple controllers on a single box?  Maybe we do only check this node so that the other ones can have a different port just for this situation?  It does seem to me to be a corner case, that normally all the ports should match.

jsancio replied:
> Hmm... what happens for the case of running multiple controllers on a single box? Maybe we do only check this node so that the other ones can have a different port just for this situation? It does seem to me to be a corner case, that normally all the ports should match.
 
This case is very common in tests, integration tests and system tests.

jsancio commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2007,12 +2008,47 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
     if (processRoles.isEmpty || processRoles.contains(BrokerRole)) {
+      require(advertisedListenerNames.nonEmpty,
+        "There must be at least one advertised listener." + (
+          if (processRoles.contains(BrokerRole)) s" Perhaps all listeners appear in ${ControllerListenerNamesProp}?" else ""))
```
Why do we need this? Doesn't Kafka advertise all of the names in `listeners` if `advertisedListenerNames` is not set?

rondagostino replied:
I added a test that exercises this path.  In particular, for the KRaft case, it is easy to get this part of the error message: `Perhaps all listeners appear in ${ControllerListenerNamesProp}?`

hachikuji replied:
Hmm, I guess this check was already implied by `advertisedListenerNames.contains(interBrokerListenerName)` below for the zk broker. Could we move it to the other `BrokerRole` handling above so that all the logic is consolidated?

rondagostino replied:
> Could we move it to the other BrokerRole handling above so that all the logic is consolidated

Done, though I only moved this check and not the two immediately following it since this is the only one that is remotely related to a KRaft config.

jsancio reviewed (commented):
Thanks for the changes @rondagostino

rondagostino committed:
check advert listeners, this voter node's port

rondagostino committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-13456

rondagostino committed:
Make KRaft broker configs in tests conform to tighter validation

rondagostino commented on core/src/main/scala/kafka/server/BrokerServer.scala:
```diff
@@ -404,10 +404,6 @@ class BrokerServer(
         config.numIoThreads, s"${SocketServer.DataPlaneMetricPrefix}RequestHandlerAvgIdlePercent",
         SocketServer.DataPlaneThreadPrefix)
 
-      if (socketServer.controlPlaneRequestChannelOpt.isDefined) {
-        throw new RuntimeException(KafkaConfig.ControlPlaneListenerNameProp + " is not " +
-          "supported when in KRaft mode.")
-      }
```
Removing since it is redundant -- we are now checking this when creating a KafkaConfig

rondagostino commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2011,12 +2012,68 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+          s"${KafkaConfig.ControllerListenerNamesProp} must not contain a value appearing in the '${KafkaConfig.ListenersProp}' configuration when running KRaft with just the broker role")
+      }
+    } else {
+      // controller listener names must be empty when not in KRaft mode
+      require(!controllerListenerNames.exists(_.nonEmpty), s"${KafkaConfig.ControllerListenerNamesProp} must be empty when not running in KRaft mode: ${controllerListenerNames.asJava.toString}")
```
Requiring controller listener names to be empty when not in KRaft mode is a new constraint that could cause existing ZK-based clusters to fail to restart.  The problem is that `controllerListeners()`, `dataPlaneListeners()`, and `advertisedListeners()` all consider the controller listeners when deciding what to return.  I believe this is technically incorrect in the ZooKeeper case because the documentation for `controller.listener.names` explicitly says "The ZK-based controller will not use this configuration."  I wasn't sure what to do here.  We could fix those 3 methods to only consider `controller.listener.names` when in KRaft mode, but it seemed more straightforward and less error-prone going forward to require that the config be empty in the ZooKeeper case -- then we can consider it all we want/wherever we want because it won't have any effect unless we are running KRaft.

I also was not sure if this requires a KIP (either a new one or a change to an existing one)

hachikuji replied:
nit: couldn't we use `$controllerListenerNames` as in the message above?

rondagostino replied:
> couldn't we use $controllerListenerNames as in the message above 

I actually will change the message above to use `${controllerListenerNames.asJava.toString}` like this one: tyhe Java version reads cleaner than the Scala version.

hachikuji replied:
Ok, at least the `toString` is redundant though.

rondagostino commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2011,12 +2012,68 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
       " Valid options are " + BrokerCompressionCodec.brokerCompressionOptions.mkString(","))
-    require(!processRoles.contains(ControllerRole) || controllerListeners.nonEmpty,
-      s"${KafkaConfig.ControllerListenerNamesProp} cannot be empty if the server has the controller role")
-
```
This check is now handled below

rondagostino commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2011,12 +2012,68 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
-
     val advertisedListenerNames = advertisedListeners.map(_.listenerName).toSet
+    if (usesSelfManagedQuorum) {
+      require(controlPlaneListenerName.isEmpty,
+        s"${KafkaConfig.ControlPlaneListenerNameProp} is not supported in KRaft mode.")
```
Used to be checked in `BrokerServer` -- now we just do it here

jsancio replied:
It looks like we have section for validating this property. Maybe move this error there:
```scala
      // validate control.plane.listener.name config
      if (controlPlaneListenerName.isDefined) {
        require(advertisedListenerNames.contains(controlPlaneListenerName.get),
          s"${KafkaConfig.ControlPlaneListenerNameProp} must be a listener name defined in ${KafkaConfig.AdvertisedListenersProp}. " +
          s"The valid options based on currently configured listeners are ${advertisedListenerNames.map(_.value).mkString(",")}")
        // controlPlaneListenerName should be different from interBrokerListenerName
        require(!controlPlaneListenerName.get.value().equals(interBrokerListenerName.value()),
          s"${KafkaConfig.ControlPlaneListenerNameProp}, when defined, should have a different value from the inter broker listener name. " +
          s"Currently they both have the value ${controlPlaneListenerName.get}")
      }
```

rondagostino commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2031,11 +2088,7 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
       s"${KafkaConfig.AdvertisedListenersProp} cannot use the nonroutable meta-address 0.0.0.0. "+
       s"Use a routable IP address.")
 
-    // Ensure controller listeners are not in the advertised listeners list
-    require(!controllerListeners.exists(advertisedListeners.contains),
-      s"${KafkaConfig.AdvertisedListenersProp} cannot contain any of ${KafkaConfig.ControllerListenerNamesProp}")
-
```
This is now checked above -- this exact check is done when running the broker role, but the check for when running just the controller role is actually now more constrained: advertised listeners must be empty in that case.

rondagostino commented on core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala:
```diff
@@ -186,34 +186,31 @@ class KafkaConfigTest {
     props.put(KafkaConfig.ListenersProp, "HOST://localhost:9091,LB://localhost:9092")
     props.put(KafkaConfig.AdvertisedListenersProp, "HOST://localhost:9091,LB://localhost:9091")
-    assertTrue(isValidKafkaConfig(props))
+    KafkaConfig.fromProps(props)
```
Simply trying to create the config is better because if the config is invalid we actually see the exception that caused it, which is much more helpful than simply seeing that `true` was expected and we got `false` instead.  I made this change in many places below as well.

rondagostino commented on core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala:
```diff
@@ -223,31 +220,26 @@ class KafkaConfigTest {
   }
 
   @Test
-  def testControllerListenerName() = {
-    val props = TestUtils.createBrokerConfig(0, TestUtils.MockZkConnect)
-    props.put(KafkaConfig.ListenersProp, "PLAINTEXT://localhost:0,CONTROLPLANE://localhost:4000,CONTROLLER://localhost:5000")
-    props.put(KafkaConfig.ListenerSecurityProtocolMapProp, "PLAINTEXT:PLAINTEXT,CONTROLPLANE:SSL,CONTROLLER:SASL_SSL")
-    props.put(KafkaConfig.AdvertisedListenersProp, "PLAINTEXT://localhost:0,CONTROLPLANE://localhost:4000")
-    props.put(KafkaConfig.ControlPlaneListenerNameProp, "CONTROLPLANE")
```
This controller listener name test included control plane listener configurations, and it was declaring the config to be using ZooKeeper instead of KRaft.  We now disallow controller.listener.names when using ZooKeeper and control.plane.listener.name when using KRaft.  So now this test does what it says: it tests controller listener name, and it sets a KRaft config and leaves out the control plane listener stuff (which is tested above anyway, in `testControlPlaneListenerName()`)

rondagostino commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -297,7 +297,7 @@ object TestUtils extends Logging {
       props.put(KafkaConfig.BrokerIdProp, nodeId.toString)
       props.put(KafkaConfig.AdvertisedListenersProp, listeners)
-      props.put(KafkaConfig.ListenersProp, listeners + ",CONTROLLER://localhost:0")
+      props.put(KafkaConfig.ListenersProp, listeners)
```
This is the same change from https://github.com/apache/kafka/pull/11511/files#diff-b8f9f9d1b191457cbdb332a3429f0ad65b50fa4cef5af8562abcfd1f177a2cfeL300.  Without adding it here we cannot get a clean build in this PR right now.

rondagostino committed:
Fix test failures in KafkaTest

rondagostino committed:
Fix test failures in ControllerApisTest

rondagostino committed:
Fix test failures in ServerTest

rondagostino committed:
Fix test failures in KafkaRaftServerTest

rondagostino committed:
Fix test failures in RaftManagerTest

rondagostino committed:
Fix test failures in BrokerLifecycleManagerTest

rondagostino committed:
Fix test failures in KafkaApisTest

rondagostino committed:
Fix test failures in ReplicaManagerConcurrencyTest

rondagostino committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-13456

rondagostino committed:
Fix test failures in KafkaMetadataLogTest

rondagostino committed:
Fix test failures in ConnectionQuotasTest

rondagostino committed:
Fix test failures in SocketServerTest

jsancio commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2018,12 +2019,68 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+            require(controllerListeners.head.port == quorumVotersPort,
+              s"Port in ${KafkaConfig.QuorumVotersProp} for this controller node (${KafkaConfig.NodeIdProp}=$nodeId, port=$quorumVotersPort) does not match the port for the first controller listener in ${KafkaConfig.ControllerListenerNamesProp} (${controllerListeners.head.listenerName.value()}, port=${controllerListeners.head.port})")
+          }
+          case _ =>
```
Can you add a comment as to why we are skipping this case? For example, I think we hit this case both when `get(nodeId)` is `null` and `UnknownAddressSpec`.

Please check this code and see if we can merge this two validation code:

```scala
        // Validate process.roles with controller.quorum.voters
        val voterIds: Set[Integer] = RaftConfig.parseVoterConnections(quorumVoters).asScala.keySet.toSet
        if (voterIds.isEmpty) {
          throw new ConfigException(s"If using ${KafkaConfig.ProcessRolesProp}, ${KafkaConfig.QuorumVotersProp} must contain a parseable set of voters.")
        } else if (processRoles.contains(ControllerRole)) {
          // Ensure that controllers use their node.id as a voter in controller.quorum.voters
          require(voterIds.contains(nodeId), s"If ${KafkaConfig.ProcessRolesProp} contains the 'controller' role, the node id $nodeId must be included in the set of voters ${KafkaConfig.QuorumVotersProp}=$voterIds  ")
        } else {
          // Ensure that the broker's node.id is not an id in controller.quorum.voters
          require(!voterIds.contains(nodeId), s"If ${KafkaConfig.ProcessRolesProp} does not contain the 'controller' role, the node id $nodeId must not be included in the set of voters ${KafkaConfig.QuorumVotersPr  op}=$voterIds")
        }
```

jsancio commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2018,12 +2019,68 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+          warn(s"${KafkaConfig.ControllerListenerNamesProp} has multiple entries; only the first will be used since ${KafkaConfig.ProcessRolesProp}=broker: $controllerListenerNames")
+        }
+        require(controllerListeners.isEmpty,
+          s"${KafkaConfig.ControllerListenerNamesProp} must not contain a value appearing in the '${KafkaConfig.ListenersProp}' configuration when running KRaft with just the broker role")
```
Are we already checking this here? https://github.com/apache/kafka/pull/11503/files#diff-cbe6a8b71b05ed22cf09d97591225b588e9fca6caaf95d3b34a43262cfd23aa6R2037-R2039

rondagostino replied:
> Are we already checking this...

This check applies to the KRaft broker-only setup and confirms that `controller.listener.names` doesn't appear in `listeners`.

The check you referred to applies to both KRaft broker setups (i.e. broker-only and co-located) and confirms that `controller.listener.names` doesn't appear in `advertised.listeners`. 

Every entry in `controller.listener.names` must appear in `listeners` for the KRaft controller setups (i.e. controller-only and co-located); we check this elsewhere.  The `advertised.listeners` config must be empty for these setups, which we also check elsewhere.  Additionally, for all broker setups (i.e. ZooKeeper and KRaft broker-only and KRaft co-located), the `advertised.listeners` config must be non-empty, it must contain the inter-broker listener, and it must be a subset of `listeners`.

I've structured the code around the setups we have to check (and will push a commit shortly to add comments that make this clear).  So we have these sections:

```
// validations for all 3 KRaft setups (co-located, controller-only, broker-only)
...
// validations for KRaft controller-only setup
...
// validations for both KRaft broker setup (i.e. broker-only and co-located)
...
// validations for both KRaft controller setups (i.e. controller-only and co-located)
...
// validations for KRaft broker-only setup
...
// validations for ZooKeeper (i.e. non-KRaft) setup
...
// validations for all broker setups (i.e. ZooKeeper and KRaft broker-only and KRaft co-located)
```

The other way to do it would be to check the 7 different possibilities for each config key/value, but I suspect that wouldn't read as well.

jsancio reviewed (commented):
Thanks for the changes @rondagostino . I think that there were some duplicate validation checks. Not sure if I was able to find all of them but can you take a look at the existing validation check and see if we can remove some duplicate code?

rondagostino committed:
Make test organization clear, adjustments based on review

cmccabe commented:
@rondagostino : I think we need to discuss the question of whether to include `controller.listener.names` in `listener.names` a bit more.

The case for including it:
1. When running in combined mode, it must be included. So if we forbid it in standalone mode, it makes it harder to switch between these modes. This would actually be the first major config divergence between standalone mode and combined mode -- currently, no other configs need to be different aside from `process.roles`.
2. Changing this is incompatible. (This isn't a deal-breaker because 3.0 / 3.1 were Preview releases for KRaft. But it is something to consider)
3. I THINK there are a bunch of cases where we assume that all the listeners are in "listeners". So some code changes are probably required. This is the least important point, of course -- we can just put in the work if we want.

The case for forbidding it:
4. It's confusing to have things in "listeners" that are not actually listeners for the current node.

A third option is to leave it as a "don't care," as it is now: you can include it, but it will be ignored, possibly with a warning. The warning option is not very enticing. We already have lots of warnings that are ignored, and it seems certain that this would become another one.

Overall I would lean towards keeping it out of "listeners," but I'm curious what you think.

Whatever we do, we need more tests of these configs... (thanks for adding more in this PR)

rondagostino commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -1992,18 +1993,6 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
         throw new ConfigException(s"Missing configuration `${KafkaConfig.NodeIdProp}` which is required " +
           s"when `process.roles` is defined (i.e. when running in KRaft mode).")
       }
-
-      // Validate process.roles with controller.quorum.voters
-      val voterIds: Set[Integer] = RaftConfig.parseVoterConnections(quorumVoters).asScala.keySet.toSet
-      if (voterIds.isEmpty) {
-        throw new ConfigException(s"If using ${KafkaConfig.ProcessRolesProp}, ${KafkaConfig.QuorumVotersProp} must contain a parseable set of voters.")
-      } else if (processRoles.contains(ControllerRole)) {
-        // Ensure that controllers use their node.id as a voter in controller.quorum.voters 
-        require(voterIds.contains(nodeId), s"If ${KafkaConfig.ProcessRolesProp} contains the 'controller' role, the node id $nodeId must be included in the set of voters ${KafkaConfig.QuorumVotersProp}=$voterIds")
-      } else {
-        // Ensure that the broker's node.id is not an id in controller.quorum.voters
-        require(!voterIds.contains(nodeId), s"If ${KafkaConfig.ProcessRolesProp} does not contain the 'controller' role, the node id $nodeId must not be included in the set of voters ${KafkaConfig.QuorumVotersProp}=$voterIds")
-      }
```
These checks are now handled below

rondagostino committed:
Don't validate voter ports against listener ports

rondagostino commented:
> I think we need to discuss the question of whether to include controller.listener.names in listener.names a bit more.

@cmccabe As discussed offline, it turns out that the current code already forbids `controller.listener.names` from appearing in `listeners` under some circumstances for the KRaft broker-only setup.

Config 1: yields **"inter.broker.listener.name must be a listener name defined in advertised.listeners."**
```
process.roles=broker
node.id=2
controller.quorum.voters=1@localhost:9093
listeners=PLAINTEXT://localhost:9092
inter.broker.listener.name=PLAINTEXT
controller.listener.names=PLAINTEXT
```

Config 2: yields **"advertised.listeners cannot contain any of controller.listener.names"**
```
process.roles=broker
node.id=2
controller.quorum.voters=1@localhost:9093
listeners=PLAINTEXT://localhost:9092
inter.broker.listener.name=PLAINTEXT
controller.listener.names=PLAINTEXT
advertised.listeners=PLAINTEXT://localhost:9092
```

Config 3: this works as it should; `controller.listener.names` does not appear in `listeners`:
```
process.roles=broker
node.id=2
controller.quorum.voters=1@localhost:9093
listeners=PLAINTEXT://localhost:9092
inter.broker.listener.name=PLAINTEXT
controller.listener.names=CONTROLLER
listener.security.protocol.map=CONTROLLER:PLAINTEXT,PLAINTEXT:PLAINTEXT
```

However, adding CONTROLLER as a listener is acceptable with the current code.  For example:

**Currently acceptable, should be rejected**
```
process.roles=broker
node.id=2
controller.quorum.voters=1@localhost:9093
listeners=PLAINTEXT://localhost:9092,CONTROLLER://localhost:9093
inter.broker.listener.name=PLAINTEXT
controller.listener.names=CONTROLLER
listener.security.protocol.map=CONTROLLER:PLAINTEXT,PLAINTEXT:PLAINTEXT
```

The inclusion of `CONTROLLER://localhost:9093` in `listeners` adds/does nothing -- in particular, the broker does **not** listen there, so really all it does is encourage confusion.

So while this is a breaking change, it is only a breaking change for certain configurations, and it is likely acceptable given that KRaft is still in Preview rather than General Availability.

It would be good to add some text about this in KIP-631.

rondagostino committed:
Add CONTROLLER:PLAINTEXT to protocol map for KRaft by default

rondagostino commented:
@cmccabe As discussed offline, I added a `CONTROLLER:PLAINTEXT` mapping to `listener.security.protocol.map` for KRaft by default as a convenience.  I added tests as well.  It would be good to include this in a KIP-631 update as well as the change to forbid `controller.listener.names` from appearing in `listeners` in the broker-only setup; the new restriction that `controller.listener.names` must be empty for the non-KRaft (i.e. ZooKeeper) setups would also be appropriate in a KIP-631 update as well.

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -748,7 +749,8 @@ object KafkaConfig {
     "INTERNAL listener, a config with name <code>listener.name.internal.ssl.keystore.location</code> would be set. " +
-    "If the config for the listener name is not set, the config will fallback to the generic config (i.e. <code>ssl.keystore.location</code>). "
+    "If the config for the listener name is not set, the config will fallback to the generic config (i.e. <code>ssl.keystore.location</code>). " +
+    "Note that in KRaft an additional default mapping CONTROLLER to PLAINTEXT is added."
```
Hmm, why do we need this? It seems strange to assume that PLAINTEXT is intended if there are multiple listeners defined. Why don't we force the user to explicitly define the controller listener?

rondagostino replied:
As discussed offline, it will be included in the KRaft **default value only** so that users don't have to set something like `listener.security.protocol.map=PLAINTEXT:PLAINTEXT:CONTROLLER:PLAINTEXT` for simple KRaft setups.  If users set anything for `listener.security.protocol.map` then the `CONTROLLER:PLAINTEXT` mapping won't be added -- users will be forced to specify everything they need if they specify anything.

hachikuji replied:
Can we generalize this comment now that the implementation has changed?

hachikuji replied:
How about this?

> Note that in KRaft, a default mapping from the listener names defined by `controller.listener.names` to PLAINTEXT is assumed if no explicit mapping is provided and no other security protocol is in use.

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2018,12 +2014,72 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+      if (addressSpecsByNodeId.isEmpty) {
+        throw new ConfigException(s"If using ${KafkaConfig.ProcessRolesProp}, ${KafkaConfig.QuorumVotersProp} must contain a parseable set of voters.")
+      }
+      require(controlPlaneListenerName.isEmpty,
```
Is it worth adding a comment to this message about the controller listener effectively superseding this config?

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2018,12 +2014,72 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+        // advertised listeners must be empty when not also running the broker role
+        require(advertisedListeners.isEmpty,
+          sourceOfAdvertisedListeners +
+            s" must only contain KRaft controller listeners from ${KafkaConfig.ControllerListenerNamesProp} when ${KafkaConfig.ProcessRolesProp}=controller")
```
Since it's a subtle point, maybe we can add some extra detail in this message. Perhaps we can say that controller have no need for advertising the listener since the endpoint is explicitly specified through controller.quorum.voters?

rondagostino replied:
> Since it's a subtle point, maybe we can add some extra detail in this message.

I tried, but it felt like it was too subtle to efficiently include -- I think it needs a paragraph, especially since you can set `advertised.listeners` to match `controller.listener.names`, which means it isn't intuitive to then say that "controllers have no need for advertising..."

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2018,12 +2014,72 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+        // when running broker role advertised listeners cannot contain controller listeners
+        require(!advertisedListenerNames.exists(aln => controllerListenerNames.contains(aln.value())),
+          sourceOfAdvertisedListeners +
+            s" must not contain KRaft controller listeners from ${KafkaConfig.ControllerListenerNamesProp} when ${KafkaConfig.ProcessRolesProp} contains the broker role")
```
Is it worth clarifying in this comment that clients only send requests to brokers?

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2018,12 +2014,72 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+        // nodeId must appear in controller.quorum.voters
+        // controller.listener.names must be non-empty
+        // every one must appear in listeners
+        require(addressSpecsByNodeId.get(nodeId) != null,
```
nit: `addressSpecsByNodeId.containsKey(nodeId)`? This would mirror the check we use below.

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -1960,9 +1962,15 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
-      ListenerName.normalised(listenerName) -> getSecurityProtocol(protocolName, KafkaConfig.ListenerSecurityProtocolMapProp)
+        ListenerName.normalised(listenerName) -> getSecurityProtocol(protocolName, KafkaConfig.ListenerSecurityProtocolMapProp)
+      }
+    val controllerListenerName = new ListenerName("CONTROLLER")
```
This is confusing. We add the "CONTROLLER" mapping even if the user has a controller listener name defined?

rondagostino replied:
> This is confusing. We add the "CONTROLLER" mapping even if the user has a controller listener name defined?

Agreed.  As per offline discussion, we will only add the `CONTROLLER:PLAINTEXT` mapping if the user has not specified a value.  Unfortunately we can't define conditional defaults, so the default value will have to remain the same `PLAINTEXT:PLAINTEXT,SSL:SSL,SASL_PLAINTEXT:SASL_PLAINTEXT,SASL_SSL:SASL_SSL`.  What I can do is check the `originals` and add the mapping when that is empty and we are using KRaft.

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2018,12 +2014,72 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+        val listenerNameValues = listeners.map(_.listenerName.value).toSet
+        require(controllerListenerNames.forall(cln => listenerNameValues.contains(cln)),
+          s"${KafkaConfig.ControllerListenerNamesProp} must only contain values appearing in the '${KafkaConfig.ListenersProp}' configuration when running the KRaft controller role")
+      } else {
```
Maybe it's just me, but I find it a bit difficult to follow these validations because of the way they are organized. I think it would be easier to follow if we organized like this:
```scala
if (processRoles == Set(BrokerRole) {
  // broker only
} else if (processRoles == Set(ControllerRole) {
  // controller only
} else {
  // mixed 
}
```
It might result in some duplication, but perhaps we could have some helpers.

rondagostino replied:
> I find it a bit difficult to follow these validations because of the way they are organized

I reorganized them as per your suggestion, and I think it is better... thanks.

rondagostino committed:
Only map CONTROLLER:PLAINTEXT when no mappings given at all

rondagostino committed:
Rearrange validations as per review

rondagostino commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2018,12 +2015,98 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+      // controller.listener.names are forbidden in listeners...
+      require(controllerListeners.isEmpty,
+        s"${KafkaConfig.ControllerListenerNamesProp} must not contain a value appearing in the '${KafkaConfig.ListenersProp}' configuration when running KRaft with just the broker role")
+      // controller.listener.names must all appear in listener.security.protocol.map
+      controllerListenerNames.filter(_.nonEmpty).foreach { name =>
+        val listenerName = ListenerName.normalised(name)
+        if (!listenerSecurityProtocolMap.contains(listenerName)) {
+          throw new ConfigException(s"Controller listener with name ${listenerName.value} defined in " +
+            s"${KafkaConfig.ControllerListenerNamesProp} not found in ${KafkaConfig.ListenerSecurityProtocolMapProp}.")
+        }
+      }
```
Requiring that `controller.listener.names` appear in `listener.security.protocol.map` for the KRaft broker-only setup is a new requirement just added in the last commit.  Without this check the following config would pass validation when in fact it is illegal:

```
process.roles=broker
node.id=1
controller.quorum.voters=2@localhost:9093
listener.security.protocol.map=PLAINTEXT:PLAINTEXT
controller.listener.names=CONTROLLER
```

For controller setups the controller listener must appear in listeners, and the security protocol map check for listeners is presumably done elsewhere.

rondagostino committed:
Another smaller rearrangement

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -1960,9 +1962,16 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+    if (usesSelfManagedQuorum && !originals.containsKey(ListenerSecurityProtocolMapProp)) {
+      // Nothing was specified explicitly, so we are using the default value;
+      // therefore, since we are using KRaft, add the CONTROLLER:PLAINTEXT mapping
+      mapValue ++ Map(new ListenerName("CONTROLLER") -> SecurityProtocol.PLAINTEXT)
```
Why do we hard-code the "CONTROLLER" listener. Shouldn't we be using the values in `controller.listener.names`?

rondagostino replied:
> Why do we hard-code the "CONTROLLER" listener. Shouldn't we be using the values in controller.listener.names?

Hmm, that's an interesting point.  We had discussed offline about adding the `CONTROLLER:PLAINTEXT` default mapping in KRaft mode, but we never considered that maybe we should default all specified controller listeners to PLAINTEXT when there is no listener security protocol map explicitly provided.  That does seem more reasonable at first glance.  @cmccabe thoughts on this?

cmccabe replied:
What if someone names their controller listener SSL ?

I suppose you could say if there the controller listener name is not the name of a security protocol, and there is no map defined, then default the listener to PLAINTEXT.

rondagostino replied:
> I suppose you could say if there the controller listener name is not the name of a security protocol, and there is no map defined, then default the listener to PLAINTEXT

Yes, definitely.  I'll go ahead and make this change.

hachikuji replied:
The default mapping here doesn't do us any good unless it corresponds to one of the values in controller.listener.names, no?

> What if someone names their controller listener SSL ?

Interestingly, the logic we use today allows for this explicitly. Here is how we currently determine the security protocol in `BrokerToControllerChannelManager`. 

```scala
    val controllerSecurityProtocol = config.listenerSecurityProtocolMap.getOrElse(controllerListenerName, SecurityProtocol.forName(controllerListenerName.value()))
```

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2018,12 +2015,104 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+
+    // validate KRaft-related configs
+    val voterAddressSpecsByNodeId = RaftConfig.parseVoterConnections(quorumVoters)
+    def validateCanParseControllerQuorumVotersForKRaft(): Unit = {
```
nit: The name is inaccurate. The parsing is done on the previous line. Maybe it should be `validateNonEmptyQuorumVoters` or something like that.

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2018,12 +2015,104 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+      require(controlPlaneListenerName.isEmpty,
+        s"${KafkaConfig.ControlPlaneListenerNameProp} is not supported in KRaft mode. KRaft uses ${KafkaConfig.ControllerListenerNamesProp} instead.")
+    }
+    def sourceOfAdvertisedListeners: String = {
```
nit: seems like this could be a `val`

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2018,12 +2015,104 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+      validateControllerQuorumVotersMustContainNodeIDForKRaftController()
+      validateControllerListenerExistsForKRaftController()
+      validateControllerListenerNamesMustAppearInListenersForKRaftController()
+      validateAdvertisedListenersNonEmptyForBroker()
```
Correct me if I'm wrong, but I think the validations here imply that mixed nodes _must_ specify `advertised.listeners`. If the user didn't provide `advertised.listeners`, then we would take the effective advertised listeners from `listeners`, which would have to include the controller listener. In other words, this would be invalid:
```
process.roles=broker,controller
controller.listener.names=CONTROLLER
listeners=DEFAULT://:9092,CONTROLLER://:9093
```

Instead, we would have to do this:
```
process.roles=broker,controller
controller.listener.names=CONTROLLER
listeners=DEFAULT://:9092,CONTROLLER://:9093
advertised.listeners=DEFAULT://:9092
```

Does that seem right? An alternative approach we could use here would be to automatically exclude the controller listeners from the advertised listeners when they are specified by `listeners`. The configuration itself gives us enough information to do the right thing.

By the way, would it make sense to rename `advertisedListeners` to `effectiveAdvertisedListeners` to make it clearer that the source might not be `advertised.listeners`?

rondagostino replied:
The config you presented as being illegal is actually acceptable assuming we also set `inter.broker.listener.name` and `listener.secuerity.protocol.map`:

```
process.roles=broker,controller
controller.listener.names=CONTROLLER
listeners=DEFAULT://:9092,CONTROLLER://:9093
# need to add these two configs to make it work
inter.broker.listener.name=DEFAULT
listener.security.protocol.map=DEFAULT:PLAINTEXT,CONTROLLER:PLAINTEXT
```

We could make it simpler if we take advantage of various default values, like this:

```
process.roles=broker,controller
controller.listener.names=CONTROLLER
listeners=PLAINTEXT://:9092,CONTROLLER://:9093
# inter.broker.listener.name and litener.security.protocol.map default values will now allow it to work
```

> An alternative approach we could use here would be to automatically exclude the controller listeners from the advertised listeners when they are specified by listeners

Yes, this is what we do:
```
  def advertisedListeners: Seq[EndPoint] = {
...
      listeners.filterNot(l => controllerListenerNames.contains(l.listenerName.value()))
  }
```

> By the way, would it make sense to rename advertisedListeners to effectiveAdvertisedListeners to make it clearer that the source might not be advertised.listeners?

Yes!  I think that makes perfect sense.

hachikuji replied:
Ah, interesting. I missed the filtering of `listeners`.

rondagostino committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-13456

rondagostino committed:
Some nit fixes, effectiveAdvertisedListeners rename

rondagostino committed:
Controller listeners default PLAINTEXT, effectiveListenerSecurityProtocolMap

rondagostino committed:
Fix Scala 2.12 compile error

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -1959,10 +1961,18 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+    if (usesSelfManagedQuorum && !originals.containsKey(ListenerSecurityProtocolMapProp)) {
+      // Nothing was specified explicitly, so we are using the default value; therefore, since we are using KRaft,
+      // add the PLAINTEXT mappings for all controller listener names that are not security protocols
+      mapValue ++ controllerListenerNames.filter(cln => cln.nonEmpty && !SecurityProtocol.values().exists(_.name.equals(cln))).map(
```
Suppose that the user has the following configuration:

```
listeners=SSL://9092,CONTROLLER://9093
controller.listener.names=CONTROLLER
```

With the change here, would we still default the CONTROLLER listener to PLAINTEXT? I can't shake the feeling that we would be better off forcing the user to be explicit about their intent. Perhaps in the one case when all of the other listeners are PLAINTEXT (as would be the case with quickstarts), then it is reasonable to also assume PLAINTEXT. But I'm inclined to be stricter if other security protocols are in use. What do you think?

rondagostino replied:
> I'm inclined to be stricter if other security protocols are in use. What do you think?

Yeah, I agree that we should only map when SSL and SASL_{SSL,PLAINTEXT} are not in use.  I've pushed a commit that implement this.

rondagostino committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-13456

rondagostino committed:
Only map to PLAINTEXT when no SSL or SASL

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -1959,10 +1962,26 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+      // and we are using KRaft.
+      // Add PLAINTEXT mappings for controller listeners as long as there is no SSL or SASL_{PLAINTEXT,SSL} in use
+      def isSslOrSasl(name: String) : Boolean = name.equals(SecurityProtocol.SSL.name) || name.equals(SecurityProtocol.SASL_SSL.name) || name.equals(SecurityProtocol.SASL_PLAINTEXT.name)
+      if (controllerListenerNames.exists(isSslOrSasl) ||
```
Checking my understanding. Is the first clause here necessary for the broker-only case in which the controller listener names are not included in `listeners`? A comment to that effect might be useful.

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2018,12 +2025,103 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+        s"${KafkaConfig.ListenersProp}"
+    def validateAdvertisedListenersDoesNotContainControllerListenersForKRaftBroker(): Unit = {
+      require(!advertisedListenerNames.exists(aln => controllerListenerNames.contains(aln.value())),
+        s"$sourceOfAdvertisedListeners must not contain KRaft controller listeners from ${KafkaConfig.ControllerListenerNamesProp} when ${KafkaConfig.ProcessRolesProp} contains the broker role because Kafka clients that send requests via advertised listeners do not send requests to KRaft controllers -- they only send requests to KRaft brokers.")
```
nit: The generality here seems misleading. We explicitly allow the controller listeners to be included among `listeners` even when `advertised.listeners` is empty. Hence I don't think it would ever be possible for `sourceOfAdvertisedListeners` to refer to `listeners` here.

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2018,12 +2025,103 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+      // controller.listener.names must all appear in listener.security.protocol.map
+      controllerListenerNames.filter(_.nonEmpty).foreach { name =>
+        val listenerName = ListenerName.normalised(name)
+        if (!effectiveListenerSecurityProtocolMap.contains(listenerName)) {
```
Probably doesn't make that much of a difference here, but we should keep in mind that `effectiveListenerSecurityProtocolMap` recomputes the listener map every time it is called. It might be worth double-checking that there are not any cases where this could be a performance concern.

rondagostino replied:
I looked around and did find a this loop in the config validation:

```
      controllerListenerNames.foreach { name =>
        val listenerName = ListenerName.normalised(name)
        if (!effectiveListenerSecurityProtocolMap.contains(listenerName)) {
          throw new ConfigException(s"Controller listener with name ${listenerName.value} defined in " +
            s"${KafkaConfig.ControllerListenerNamesProp} not found in ${KafkaConfig.ListenerSecurityProtocolMapProp}.")
        }
      }
```

 As you pointed out, it probably doesn't matter given the small sizes.  However, I can't think of a reason to not make it a `val` and only calculate it once, so I made that change.  Let me know if I'm missing something and this shouldn't be done.

rondagostino replied:
> I can't think of a reason to not make it a val and only calculate it once, so I made that change

Had to revert since DynamicConnectionQuotaTest was failing, but I believe that is fine.

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2018,12 +2025,103 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+      require(controllerListeners.isEmpty,
+        s"${KafkaConfig.ControllerListenerNamesProp} must not contain a value appearing in the '${KafkaConfig.ListenersProp}' configuration when running KRaft with just the broker role")
+      // controller.listener.names must all appear in listener.security.protocol.map
+      controllerListenerNames.filter(_.nonEmpty).foreach { name =>
```
Why do we allow empty controller listener names?

rondagostino replied:
> Why do we allow empty controller listener names?

No reason -- we shouldn't.  I'll fix it.

hachikuji commented on core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala:
```diff
@@ -1000,46 +1139,78 @@ class KafkaConfigTest {
-    props.put(KafkaConfig.ListenersProp, listeners)
-    props.put(KafkaConfig.AdvertisedListenersProp, "PLAINTEXT://A:9092,SSL://B:9093")
+  @Test
+  def assertDistinctControllerAndAdvertisedListenersAllowedForKRaftBroker(): Unit = {
```
nit: it would be nice to keep the convention of prefixing test cases with "test"

hachikuji commented on core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala:
```diff
@@ -1000,46 +1139,78 @@ class KafkaConfigTest {
   @Test
-  def assertEvenOneControllerListenerCannotBeAdvertised(): Unit = {
-    val props = TestUtils.createBrokerConfig(0, TestUtils.MockZkConnect, port = TestUtils.MockZkPort)
+  def assertAdvertisedListenersDisallowedForKRaftControllerOnlyRole(): Unit = {
```
I might have missed it, but do we have a test for the scenario where SSL/SASL are in use and no controller listener security mapping is defined?

rondagostino replied:
> do we have a test for the scenario where SSL/SASL are in use and no controller listener security mapping is defined 

Yes, testControllerListenerNameMapsToPlaintextByDefaultForKRaft() has this case.

rondagostino committed:
nits, cleanups

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -748,7 +750,9 @@ object KafkaConfig {
     "INTERNAL listener, a config with name <code>listener.name.internal.ssl.keystore.location</code> would be set. " +
-    "If the config for the listener name is not set, the config will fallback to the generic config (i.e. <code>ssl.keystore.location</code>). "
+    "If the config for the listener name is not set, the config will fallback to the generic config (i.e. <code>ssl.keystore.location</code>). " +
+    "Note that in KRaft a default mapping from the listener names defined by controller.listener.names to PLAINTEXT " +
```
nit: consistent with the rest of the comment, can we use `<code>` blocks around `controller.listener.names`?

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -1887,15 +1891,21 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
-    Option(getString(KafkaConfig.ControllerListenerNamesProp)).getOrElse("").split(",")
+  def controllerListenerNames: Seq[String] = {
+    val value = Option(getString(KafkaConfig.ControllerListenerNamesProp)).getOrElse("")
+      if (value.isEmpty) {
```
nit: this looks misaligned

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -1959,10 +1969,28 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
 
-  def listenerSecurityProtocolMap: Map[ListenerName, SecurityProtocol] = {
-    getMap(KafkaConfig.ListenerSecurityProtocolMapProp, getString(KafkaConfig.ListenerSecurityProtocolMapProp))
+  def effectiveListenerSecurityProtocolMap: Map[ListenerName, SecurityProtocol] = {
```
One more thing to check is that the validations in `DynamicListenerConfig` make sense in the context of these changes.

rondagostino replied:
> One more thing to check is that the validations in DynamicListenerConfig make sense in the context of these changes 

I just push a commit fixing all of the minor issues, I'll look into this possible issue later today.

rondagostino replied:
> validations in DynamicListenerConfig make sense in the context of these changes?

We do have this code, which means we currently don't have to worry about this for KRaft:

https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/DynamicBrokerConfig.scala#L909-L911

However, we will support this eventually, so let's put that aside for the moment.

We do have these checks, which continue to make sense:

```
      if (immutableListenerConfigs(newConfig, listenerName.configPrefix) != immutableListenerConfigs(oldConfig, listenerName.configPrefix))
        throw new ConfigException(s"Configs cannot be updated dynamically for existing listener $listenerName, " +
          "restart broker or create a new listener for update")
      if (oldConfig.effectiveListenerSecurityProtocolMap(listenerName) != newConfig.effectiveListenerSecurityProtocolMap(listenerName))
        throw new ConfigException(s"Security protocol cannot be updated for existing listener $listenerName")
```

One thing that could occur is that we add a new SSL listener that is the first one that is non-PLAINTEXT and then all of a sudden we have removed any PLAINTEXT default values in the listener security protocol map, which could cause the new config to fail.  But that's fine.  And this would never happen in a non-toy cluster since those generally have at least one secured/non-PLAINTEXT listener anyway.

Another possibility is that we might impact existing controller listeners, but that impact would have to be checked for when we implement this dynamic reconfiguration path for KRaft.

My take on this is that the existing changes are okay in this context.

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -1959,10 +1969,28 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+      // Nothing was specified explicitly for listener.security.protocol.map, so we are using the default value,
+      // and we are using KRaft.
+      // Add PLAINTEXT mappings for controller listeners as long as there is no SSL or SASL_{PLAINTEXT,SSL} in use
+      def isSslOrSasl(name: String) : Boolean = name.equals(SecurityProtocol.SSL.name) || name.equals(SecurityProtocol.SASL_SSL.name) || name.equals(SecurityProtocol.SASL_PLAINTEXT.name)
```
nit: remove space before colon

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2018,12 +2034,103 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+        val listenerName = ListenerName.normalised(name)
+        if (!effectiveListenerSecurityProtocolMap.contains(listenerName)) {
+          throw new ConfigException(s"Controller listener with name ${listenerName.value} defined in " +
+            s"${KafkaConfig.ControllerListenerNamesProp} not found in ${KafkaConfig.ListenerSecurityProtocolMapProp}.")
```
It might be nice to add in parenthesis the following:
`(an explicit security mapping for each controller listener is required if ${KafkaConfig.ListenerSecurityProtocolMapProp} is non-empty, or if there are security protocols other than PLAINTEXT in use)`

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -2018,12 +2034,103 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+      require(!advertisedListenerNames.exists(aln => controllerListenerNames.contains(aln.value())),
+        s"The advertised.listeners config must not contain KRaft controller listeners from ${KafkaConfig.ControllerListenerNamesProp} when ${KafkaConfig.ProcessRolesProp} contains the broker role because Kafka clients that send requests via advertised listeners do not send requests to KRaft controllers -- they only send requests to KRaft brokers.")
+    }
+    def validateControllerQuorumVotersMustContainNodeIDForKRaftController(): Unit = {
```
nit: everywhere else, we use `NodeId` instead of `NodeID`

hachikuji reviewed (commented):
LGTM overall. Thanks for the iterations. I left a couple nits and a question about `DynamicListenerConfig` which I haven't had time to look into yet.

rondagostino committed:
nits

rondagostino committed:
Fix failing test

hachikuji reviewed (approved):
LGTM!
